libs/sfm/bundle_adjustment.cc

```diff
@@ -629,17 +662,28 @@ void
 BundleAdjustment::update_camera (Camera const& cam,
     double const* update, Camera* out)
 {
-    out->focal_length = cam.focal_length + update[0];
-    out->distortion[0] = cam.distortion[0] + update[1];
-    out->distortion[1] = cam.distortion[1] + update[2];
-    out->translation[0] = cam.translation[0] + update[3];
-    out->translation[1] = cam.translation[1] + update[4];
-    out->translation[2] = cam.translation[2] + update[5];
+    if (opts.fixed_intrinsics)
+    {
+        out->focal_length = cam.focal_length;
+        out->distortion[0] = cam.distortion[0];
+        out->distortion[1] = cam.distortion[1];
+    }
+    else
+    {
+        out->focal_length = cam.focal_length + update[0];
+        out->distortion[0] = cam.distortion[0] + update[1];
+        out->distortion[1] = cam.distortion[1] + update[2];
+    }
+
+    int const offset = this->opts.fixed_intrinsics ? 3 : 0;
+    out->translation[0] = cam.translation[0] + update[3 - offset];
+    out->translation[1] = cam.translation[1] + update[4 - offset];
+    out->translation[2] = cam.translation[2] + update[5 - offset];
 
     double rot_orig[9];
     std::copy(cam.rotation, cam.rotation + 9, rot_orig);
     double rot_update[9];
-    this->rodrigues_to_matrix(update + 6, rot_update);
+    this->rodrigues_to_matrix(update + 6 - offset, rot_update);
```

Comment:
How about setting `offset` to either 6 or 3 based on `this->opts.fixed_intrinsics` and just do `update + offset` here?